sound/soc/sof/intel/cnl.c

```diff
@@ -150,14 +151,77 @@ static void cnl_ipc_dsp_done(struct snd_sof_dev *sdev)
 				CNL_DSP_REG_HIPCCTL_DONE);
 }
 
+static bool cnl_pm_gate_compress(struct snd_sof_ipc_msg *msg,
+				 u32 *dr, u32 *dd)
+{
+	struct sof_ipc_pm_gate *pm_gate;
+
+	if (msg->header == (SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_GATE)) {
+		pm_gate = msg->msg_data;
+
+		/* send the compact message via the primary register */
+		*dr = HDA_IPC_MSG_COMPACT | HDA_IPC_PM_GATE;
+
+		/* send payload via the extended data register */
+		*dd = pm_gate->flags;
+
+		return true;
+	}
+
+	return false;
+}
+
+static const struct hda_ipc compact_ipc[] = {
+	{.header = SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_GATE,
+	 .compress = cnl_pm_gate_compress},
+};
+
+static bool cnl_is_ipc_compact(struct snd_sof_ipc_msg *msg)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(compact_ipc); i++)
+		if (msg->header == compact_ipc[i].header)
+			return true;
+
+	return false;
+}
+
+static bool cnl_compact_ipc_compress(struct snd_sof_ipc_msg *msg,
+				     u32 *dr, u32 *dd)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(compact_ipc); i++)
+		if (msg->header == compact_ipc[i].header)
+			return compact_ipc[i].compress(msg, dr, dd);
+
+	return false;
+}
+
 static int cnl_ipc_send_msg(struct snd_sof_dev *sdev,
 			    struct snd_sof_ipc_msg *msg)
 {
-	/* send the message */
-	sof_mailbox_write(sdev, sdev->host_box.offset, msg->msg_data,
-			  msg->msg_size);
-	snd_sof_dsp_write(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDR,
-			  CNL_DSP_REG_HIPCIDR_BUSY);
+	u32 dr = 0;
+	u32 dd = 0;
+
+	if (cnl_is_ipc_compact(msg)) {
+		/* first, configure the IPC registers */
+		if (!cnl_compact_ipc_compress(msg, &dr, &dd))
+			return -EINVAL;
+
+		/* then, send the message via IPC registers */
+		snd_sof_dsp_write(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDD,
+				  dd);
+		snd_sof_dsp_write(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDR,
+				  CNL_DSP_REG_HIPCIDR_BUSY | dr);
```

Comment:
so we send the message over IPC registers, but is there a reply?
I remember talking with @mmaka1 that we could have replies over IPC registers as well.

Reply:
we also have register HIPCTDD, but haven't implemented reading of them(FW actually not fill them in Marcin's code yet), we only get the IPC interrupt and check the header from HIPCTDR only.

Reply:
@keyonjie I think we could simplify this logic like this (for example):
```
static int cnl_get_ipc_pm_type(struct snd_sof_ipc_msg *msg)
{
	struct sof_ipc_cmd_hdr *hdr = msg->msg_data;
	u32 glb;
	u32 type;

	glb = hdr->cmd & SOF_GLB_TYPE_MASK;
	type = hdr->cmd & SOF_CMD_TYPE_MASK;

	if (glb & SOF_IPC_GLB_PM_MSG) {
		switch(type) {
		case SOF_IPC_PM_GATE:
			return HDA_IPC_PM_GATE;
		default:
			return -EINVAL;
		}		
	}

	return -EINVAL;

}

static int cnl_ipc_send_msg(struct snd_sof_dev *sdev,
			    struct snd_sof_ipc_msg *msg)
{
	u32 dr = 0;
	u32 dd = 0;
	int pm_ipc_type;

	pm_ipc_type = cnl_get_ipc_pm_type(msg);
	if (ipc_type > 0) {
		/* set compact ipc type */
		*dr = HDA_IPC_MSG_COMPACT | ipc_type;

		/* set payload via the extended data register */
		*dd = msg->msg_data;

		/* send the message via IPC registers */
		snd_sof_dsp_write(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDD,
				  dd);
		snd_sof_dsp_write(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDR,
				  CNL_DSP_REG_HIPCIDR_BUSY | dr);
	} else {
		/* send the message via mailbox */
		sof_mailbox_write(sdev, sdev->host_box.offset, msg->msg_data,
				  msg->msg_size);
		snd_sof_dsp_write(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDR,
				  CNL_DSP_REG_HIPCIDR_BUSY);
	}

	return 0;
}
```

Reply:
> @keyonjie I think we could simplify this logic like this (for example):
> 
> ```
> static int cnl_get_ipc_pm_type(struct snd_sof_ipc_msg *msg)

Is this function designed to handle all GLB_PM_MSG message or only for PM_GATE?
If it is the former case, we should take care of CTX_SAVE/RESTORE also, where the msg_data is actually quite large:
struct sof_ipc_pm_ctx {
	struct sof_ipc_cmd_hdr hdr;
	struct sof_ipc_host_buffer buffer;
	uint32_t num_elems;
	uint32_t size;

	/* reserved for future use */
	uint32_t reserved[8];

	struct sof_ipc_pm_ctx_elem elems[];
} __packed;

If it is the latter case, the we should rename the function from cnl_get_ipc_pm_type to cnl_is_ipc_pm_gate ?

> {
> 	struct sof_ipc_cmd_hdr *hdr = msg->msg_data;
> 	u32 glb;
> 	u32 type;
> 
> 	glb = hdr->cmd & SOF_GLB_TYPE_MASK;
> 	type = hdr->cmd & SOF_CMD_TYPE_MASK;
> 
> 	if (glb & SOF_IPC_GLB_PM_MSG) {
> 		switch(type) {
> 		case SOF_IPC_PM_GATE:
> 			return HDA_IPC_PM_GATE;
> 		default:
> 			return -EINVAL;
> 		}		
> 	}
> 
> 	return -EINVAL;
> 
> }
> 
> static int cnl_ipc_send_msg(struct snd_sof_dev *sdev,
> 			    struct snd_sof_ipc_msg *msg)
> {
> 	u32 dr = 0;
> 	u32 dd = 0;
> 	int pm_ipc_type;
> 
> 	pm_ipc_type = cnl_get_ipc_pm_type(msg);
> 	if (pm_ipc_type > 0) {
> 		/* set compact ipc type */
> 		*dr = HDA_IPC_MSG_COMPACT | pm_ipc_type;
> 
> 		/* set payload via the extended data register */
> 		*dd = msg->msg_data;

msg_data is a '* void', *dd is a 'u32', do you mean only take the first u32 of the buffer?

> 
> 		/* send the message via IPC registers */
> 		snd_sof_dsp_write(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDD,
> 				  dd);
> 		snd_sof_dsp_write(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDR,
> 				  CNL_DSP_REG_HIPCIDR_BUSY | dr);
> 	} else {
> 		/* send the message via mailbox */
> 		sof_mailbox_write(sdev, sdev->host_box.offset, msg->msg_data,
> 				  msg->msg_size);
> 		snd_sof_dsp_write(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDR,
> 				  CNL_DSP_REG_HIPCIDR_BUSY);
> 	}
> 
> 	return 0;
> }
> ```

Reply:
@keyonjie no please rename it to cnl_get_ipc_pm_compact_type() to be more precise

Reply:
> msg_data is a '* void', *dd is a 'u32', do you mean only take the first u32 of the buffer?
> 
> >

yes please do the needful here. my code was only an example.

Reply:
the bottom line is the check for whether it is a compact IPC and what type can be done in one go.

Reply:
> the bottom line is the check for whether it is a compact IPC and what type can be done in one go.

then we have much easier way to do that if we don't need it to be extendable.

will submit update soon.

Reply:
And yes, for every change, I need some more minutes to run testing before submit it.